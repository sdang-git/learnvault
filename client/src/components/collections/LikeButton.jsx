import React, { useState } from 'react';

import './LikeButton.css';

const LikeButton = ({ id, loggedInUser, likes }) => {
  function likeButtonClick(eventId, userId) {
    const payload = { id: userId, collectionId: id };

    fetch(`/api/collections/like/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    })
      .then((response) => response.json())
      .then((data) => {
        console.log('Success:', data);
      })
      .catch((error) => {
        console.error('Error:', error);
      });
  }

  return (
    <>
      <span
        onClick={() => likeButtonClick(id, loggedInUser, likes)}
        onKeyPress={() => likeButtonClick(id, loggedInUser, likes)}
        type="button"
        className="button-like"
        role="button"
        tabIndex={0}
      >
        <i className="far fa-thumbs-up" />
        <small>{likes.length}</small>
      </span>
    </>
  );
};

export default LikeButton;
