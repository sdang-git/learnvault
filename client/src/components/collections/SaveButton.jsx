import React from 'react';

const SaveButton = ({ id, loggedInUser }) => {
  function likeButtonClick(eventId, userId) {
    const payload = { id: userId, collectionId: id };

    fetch(`/api/collections/save/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
    })
      .then((response) => response.json())
      .then((data) => {
        console.log('Success:', data);
      })
      .catch((error) => {
        console.error('Error:', error);
      });
  }

  return (
    <span
      onClick={() => likeButtonClick(id, loggedInUser)}
      onKeyPress={() => likeButtonClick(id, loggedInUser)}
      role="button"
      tabIndex={0}
      type="button"
      className="button-like"
    >
      <i className="far fa-star" />
    &nbsp; Save Collection
    </span>
  );
};

export default SaveButton;
