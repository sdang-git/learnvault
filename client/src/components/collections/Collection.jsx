import React from 'react';
import { Link } from 'react-router-dom';

import LikeButton from './LikeButton';
import SaveButton from './SaveButton';

import './Collection.css';

const Collection = ({
  id, title, description, author, loggedInUser, likes,
}) => (
  <div key={id} className="collection">
    <h1 className="collection__title">{title}</h1>
    <h3>{description}</h3>

    <div className="creator">
      <div className="creator__label">Creator:</div>
      <div className="creator__author">{author}</div>
    </div>
    <div>
      <Link to={`/collections/${id}`} className="collection__button">
        View Collection
      </Link>
    </div>
    {loggedInUser ? (
      <div>
        <br />
        <LikeButton loggedInUser={loggedInUser} likes={likes} id={id} />
        <SaveButton loggedInUser={loggedInUser} id={id} />
      </div>
    ) : (
      <div>
        <br />
        <Link to="/register">Register</Link>
        &nbsp;or&nbsp;
        <Link to="/login">Login</Link>
        &nbsp;to save this collection
      </div>
    )}
  </div>
);

export default Collection;
