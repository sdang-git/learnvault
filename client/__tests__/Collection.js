/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import LikeButton from '../src/components/collections/LikeButton';
import SaveButton from '../src/components/collections/SaveButton';
import Collection from '../src/components/collections/Collection';

configure({ adapter: new Adapter() });

describe('Collection tests', () => {
  describe('Collection', () => {
    let wrapper;
    const testProps = {
      key: 544,
      id: 4354,
      title: 'Collection',
      description: 'buncha stuff',
      author: 'Veronica',
      loggedInUser: 'Sarge',
      likes: ['5ef2b8c3d5973033a191aea2', '5ef3f1798a8800471b987bbe', '5ef7840ba4aca16282b26299'],
    };

    beforeAll(() => {
      wrapper = shallow(
        <Collection
          key={testProps.key}
          id={testProps.id}
          title={testProps.title}
          description={testProps.description}
          author={testProps.author}
          loggedInUser={testProps.loggedInUser}
          likes={testProps.likes}
        />,
      );
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders an h1 element that displays collection title', () => {
      expect(wrapper.find('h1').text()).toMatch(testProps.title);
    });

    it('renders an h3 element that displays collection description', () => {
      expect(wrapper.find('h3').text()).toMatch(testProps.description);
    });

    it('displays div with name of author', () => {
      expect(wrapper.find('div').first().text()).toMatch(testProps.author);
    });

    it('displays a Link with view the collection', () => {
      expect(wrapper.find('Link').text()).toMatch('View Collection');
    });

    it('displays Like and Save buttons if logged in', () => {
      expect(wrapper.find('LikeButton')).toHaveLength(1);
      expect(wrapper.find('SaveButton')).toHaveLength(1);
    });

    // alternate way to actually verify the components are being rendered
    // uses package: jest-enzyme
    it('renders the LikeButton and SaveButton components if logged in', () => {
      expect(wrapper).toContainReact(<LikeButton loggedInUser={testProps.loggedInUser} likes={testProps.likes} id={testProps.id} />);
      expect(wrapper).toContainReact(<SaveButton loggedInUser={testProps.loggedInUser} id={testProps.id} />);
    });

    it('displays message to register or login if not logged in', () => {
      wrapper = shallow(
        <Collection
          key={testProps.key}
          id={testProps.id}
          title={testProps.title}
          description={testProps.description}
          author={testProps.author}
          loggedInUser={null}
          likes={testProps.likes}
        />,
      );
      expect(wrapper.text()).toMatch('Register');
      expect(wrapper.text()).toMatch('Login');
    });
  });
});
