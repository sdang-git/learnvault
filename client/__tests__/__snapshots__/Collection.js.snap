// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection tests Collection renders correctly 1`] = `
<div
  className="collection"
  key="4354"
>
  <h1
    className="collection__title"
  >
    Collection
  </h1>
  <h3>
    buncha stuff
  </h3>
  <div
    className="creator"
  >
    <div
      className="creator__label"
    >
      Creator:
    </div>
    <div
      className="creator__author"
    >
      Veronica
    </div>
  </div>
  <div>
    <Link
      className="collection__button"
      to="/collections/4354"
    >
      View Collection
    </Link>
  </div>
  <div>
    <br />
    <LikeButton
      id={4354}
      likes={
        Array [
          "5ef2b8c3d5973033a191aea2",
          "5ef3f1798a8800471b987bbe",
          "5ef7840ba4aca16282b26299",
        ]
      }
      loggedInUser="Sarge"
    />
    <SaveButton
      id={4354}
      loggedInUser="Sarge"
    />
  </div>
</div>
`;
